rust/pattern/variable/builder/thing_variable_builder.rs

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+use crate::pattern::*;
+use std::fmt::Debug;
+
+pub trait ThingVariableBuilder: Sized {
+    fn has<T: TryInto<Value>>(self, type_name: impl Into<String>, value: T) -> BoundVariable
+    where
+        <T as TryInto<Value>>::Error: Debug,
+    {
+        self.constrain_thing(
+            match value.try_into().unwrap() {
+                Value::Variable(variable) => HasConstraint::from_typed_variable(
+                    type_name.into(),
+                    *variable,
+                ),
+                value => HasConstraint::from_value(
+                    type_name.into(),
+                    ValueConstraint::new(Predicate::Eq, value),
+                ),
+            }
+            .into_thing_constraint(),
+        ).into_bound_variable()
+    }
+
+    fn isa(self, isa: impl Into<IsaConstraint>) -> BoundVariable {
+        self.constrain_thing(isa.into().into_thing_constraint()).into_bound_variable()
+    }
+
+    fn eq(self, value: impl Into<Value>) -> BoundVariable {
+        self.constrain_thing(
+            ValueConstraint::new(Predicate::Eq, value.into()).into_thing_constraint(),
+        ).into_bound_variable()
+    }
+
+    fn constrain_thing(self, constraint: ThingConstraint) -> ThingVariable;
```

Comment:
`constrain_thing` & `constrain_type` — I'm not sure these are ever used when we don't know exactly what type the constraint is. It'd probably be wise to have `constrain_isa()` etc instead.